### PR TITLE
Redo the test seeding system

### PR DESCRIPTION
**Pull request by bcm-at-zama**
The seeding system is completely changed and simplified: now, you launch a test with `pytest`, which is going to use a randomly seed (either set by you or picked randomly), and this randomly seed is all what you need: the seeds used for all tests are going to be derived from this master seed, whatever the order of the tests. Such that, if there is a crash, you don't need a print of some subseed to reproduce.

More precisely: the used seed for a test of a function `f_i` (eg, test_compute_bits_precision) on a configuration `c_j` (eg, x0-8) of a test file `t_k` (eg, tests/common/test_utils.py) is computed as some `hash(f_i, c_j, t_k, randomly-seed)`

Or, in other words, if for some randomly-seed, a test crash for `(f_i, c_j, t_k)`, you can reproduce the same seed by directly calling: `python -m pytest t_k::f_i[c_j] --randomly-seed=XXX`, eg `python -m pytest -xsvv "tests/common/test_utils.py::test_compute_bits_precision[x0-8]" --randomly-seed=1508248188`.

### CODE REVIEW

**Comment by bcm-at-zama** [comment]
I did that for our issue in the CI. However, I was not able to reproduce this exact issue, so I wouldn't be able if it perfectly works. However, I'm pretty confident it's the real seed (and it will be nice in the future to try the system and check it).

**Comment by bcm-at-zama** [comment]
Hum, the fact that it says 

--forcing_random_seed 12679042460344433083 --randomly-dont-reset-seed

too much is not a good sign, maybe I should have another look.

**Comment by bcm-at-zama** [comment]
It will never work like this, random's are picked in the test so we don't simulate this by aborting at the beginning. We would need a major change in the seeding system to make it work as expected, let me see

File: conftest.py [hunk 1]
```diff
@@ -49,15 +50,6 @@ def pytest_addoption(parser):
         help="To dump pytest-cov term report to a text file.",
     )
 
-    parser.addoption(
-        "--forcing_random_seed",
```
Comment (bcm-at-zama) [inline conftest.py:53]: --forcing_random_seed is no more needed! the master seed is the randomly seed, which is unique and always printed at the beginning

File: conftest.py [hunk 2]
```diff
@@ -176,37 +168,61 @@ def function_to_seed_torch(seed):
...
+    assert (
+        absolute_path.count("concrete-ml/test") == 1
+    ), f"the absolute path is not as expected {absolute_path=}"
+
+    relative_file_path = absolute_path[
+        absolute_path.find("concrete-ml/test") + len("concrete-ml/") :
+    ]
+
+    # Derive the sub_seed from the randomly_seed and the test name
+    derivation_string = f"{relative_file_path} # {str(request.node.name)} # {randomly_seed}"
```
Comment (bcm-at-zama) [inline conftest.py:191]: this is the derivation of the subseed: for a test of a function f_i (eg, test_compute_bits_precision) on a configuration c_j (eg, x0-8) of a test file t_k (eg, tests/common/test_utils.py) is computed as some hash(f_i, c_j, t_k, randomly-seed)

File: conftest.py [hunk 3]
```diff
@@ -176,37 +168,61 @@ def function_to_seed_torch(seed):
...
+    seed = sub_seed
 
-    # Python
-    random.seed(seed)
-    print("\nForcing seed to random.seed to ", seed)
     print(
-        f"\nRelaunch the tests with --forcing_random_seed {seed} "
-        + "--randomly-dont-reset-seed to reproduce. Remark that adding --randomly-seed=... "
-        + "is needed when the testcase uses randoms in pytest parameters"
+        f"\nRelaunch the tests with --randomly_seed {randomly_seed} "
```
Comment (bcm-at-zama) [inline conftest.py:206]: repro directly a subtest by:
- forcing --randomly_seed
- setting the subtest you want
and that's it

File: script/make_utils/check_pytest_determinism.sh [hunk 4]
```diff
@@ -40,32 +40,25 @@ set -e
...
 echo "Successful determinism check"
 
-# Now, check --forcing_random_seed, i.e., check that one can reproduce conditions of a bug in a single file
-# and test without having to relaunch the full pytest, by just picking the right --forcing_random_seed
+# Now, check that one can reproduce conditions of a bug in a single file
+# and test without having to relaunch the full pytest
 
 # All lines that start with "tests/seeding/test_seeding.py::" represent a test that passed
 LIST_FILES=$(grep "tests/seeding/test_seeding.py::" "${OUTPUT_DIRECTORY}/one.txt")
-LIST_SEED=()
```
Comment (bcm-at-zama) [inline script/make_utils/check_pytest_determinism.sh:48]: no more need to find --forcing_random_seed in the log files! it was especially boring since sometimes, they were not there if a crash occurred.

File: conftest.py [hunk 5]
```diff
@@ -176,37 +168,61 @@ def function_to_seed_torch(seed):
...
-def autoseeding_of_everything(record_property, request):
+def autoseeding_of_everything(request):
     """Seed everything we can, for determinism."""
-    main_seed = request.config.getoption("--forcing_random_seed", default=None)
+    randomly_seed = request.config.getoption("--randomly-seed", default=None)
 
-    if main_seed is None:
-        main_seed = random.randint(0, 2**64 - 1)
+    if randomly_seed is None:
+        raise ValueError("why isn't --randomly-seed set?")
```
Comment (RomanBredehoft) [inline conftest.py:176]: would prefer a message like `"Please set --randomly-seed set with an integer"`

File: conftest.py [hunk 6]
```diff
@@ -176,37 +168,61 @@ def function_to_seed_torch(seed):
...
 
-    seed = main_seed
-    record_property("main seed", main_seed)
+    # We need to find the relative file path of the test file. It does not look native with request,
+    # so we recompute it.
+    absolute_path = str(request.fspath)
+
+    assert (
+        absolute_path.count("concrete-ml/test") == 1
```
Comment (RomanBredehoft) [inline conftest.py:183]: not sure what this is, and the associated error remains pretty vague !

File: conftest.py [hunk 7]
```diff
@@ -176,37 +168,61 @@ def function_to_seed_torch(seed):
...
+    ), f"the absolute path is not as expected {absolute_path=}"
+
+    relative_file_path = absolute_path[
+        absolute_path.find("concrete-ml/test") + len("concrete-ml/") :
```
Comment (RomanBredehoft) [inline conftest.py:187]: not sure what this does either

File: conftest.py (hunk 5 above)
Comment (bcm-at-zama) [reply conftest.py:176]: But they don't have to set this themselves, it's done by the system. In other words, it's normally impossible to raise this error

Comment (bcm-at-zama) [reply conftest.py:176]: I can change the sentence if you want, but not to what you mentioned because it's misleading

File: conftest.py (hunk 2 above)
Comment (RomanBredehoft) [reply conftest.py:191]: why not indeed ! is `request.node.name` the actual test + config ? like `tests/common/test_utils.py::test_compute_bits_precision[True-False-5]` ?

File: conftest.py (hunk 6 above)
Comment (bcm-at-zama) [reply conftest.py:183]: I have an absolute path, eg `/Users/benoitchevalliermames/Documents/Zama/Git/concrete-ml/tests/common/test_utils.py`
I want to compute `tests/common/test_utils.py` from it
It will work, as soon as the guy making the cloning didn't set a crazy path like

`/something/concrete-ml/tests/something-else/concrete-ml/tests/`

in which case I want to abort because it's not good.

File: conftest.py [hunk 8]
```diff
@@ -176,37 +168,61 @@ def function_to_seed_torch(seed):
...
+
+    # Derive the sub_seed from the randomly_seed and the test name
+    derivation_string = f"{relative_file_path} # {str(request.node.name)} # {randomly_seed}"
+
+    hash_object = hashlib.sha256()
+    hash_object.update(b"{derivation_string}")
+    hash_object.digest()
+    hash_value = hash_object.hexdigest()
+
+    sub_seed = int(hash_value, 16) % 2**64
```
Comment (RomanBredehoft) [inline conftest.py:198]: why is this needed ?

File: conftest.py (hunk 6 above)
Comment (bcm-at-zama) [reply conftest.py:183]: If you really want, I can make a change to find the _last_ occurence of `concrete-ml/tests`, instead of the first one. But except if the guys want to bother us, never the assert will fail

File: conftest.py (hunk 7 above)
Comment (bcm-at-zama) [reply conftest.py:187]: explained in

+    # We need to find the relative file path of the test file. It does not look native with request,
+    # so we recompute it.

two lines before. Not sufficient, or a better way to do it?

File: conftest.py (hunk 3 above)
Comment (RomanBredehoft) [reply conftest.py:206]: what do you mean by "setting the subtest you want" ? isn't running `pytest tests/common/test_utils.py::test_compute_bits_precision[True-False-5] --randomly_seed 78975873 --randomly-dont-reset-seed` enough for example ?

File: conftest.py (hunk 8 above)
Comment (bcm-at-zama) [reply conftest.py:198]: why _what_ is needed? which part?

File: conftest.py [hunk 9]
```diff
@@ -176,37 +168,61 @@ def function_to_seed_torch(seed):
...
+        f"\nRelaunch the tests with --randomly_seed {randomly_seed} "
+        + "--randomly-dont-reset-seed to reproduce."
     )
     print(
         "Remark that potentially, any option used in the pytest call may have an impact so in "
         + "case of problem to reproduce, you may want to have a look to `make pytest` options"
     )
 
+    # Python
+    random.seed(seed)
```
Comment (RomanBredehoft) [inline conftest.py:215]: are you sure numpy does not already do that ?

File: deps_licenses/licenses_mac_intel_user.txt.md5 [hunk 10]
```diff
@@ -1 +1 @@
-f709427468f5be4ee6836603495fea72
+a923947bfb17b658ab8efe61d5cafe96
```
Comment (RomanBredehoft) [inline deps_licenses/licenses_mac_intel_user.txt.md5:1]: you should run the licence update action and target your branch to also update linux + M1 licenses : https://github.com/zama-ai/concrete-ml/actions/workflows/update_licenses.yaml

File: conftest.py (hunk 2 above)
Comment (bcm-at-zama) [reply conftest.py:191]: if we have two files t_k in two different directories with the same filename: it would give the same hash and so subseed. I wanted to keep the "all the tests have different seed" properties, so I included the path as well. But I don't want absolute path, such that, whatever our config (eg, mine or yours), the seeds are the same. In other words, if you clone CML in a dir R and I do it in a dir B, our seeds are the same

File: conftest.py (hunk 3 above)
Comment (bcm-at-zama) [reply conftest.py:206]: Now with the changes, yes `pytest tests/common/test_utils.py::test_compute_bits_precision[True-False-5] --randomly_seed 78975873 --randomly-dont-reset-see` is enough but that's exactly the change. Before, it was NOT sufficient, we have to use `--forcing_random_seed ` option

**Review by RomanBredehoft — CHANGES_REQUESTED** [review]
thanks for that ! I made a few comments ! Also : 
- I would add a few comments explaining a bit more what you are doing (notably where I asked some questions)
- could you also update the licenses using the dedicated action ? 
- is there anything to change in the makefile as well ?

File: conftest.py (hunk 3 above)
Comment (bcm-at-zama) [reply conftest.py:206]: Tried to re-explain what it does in this thread

File: conftest.py (hunk 9 above)
Comment (bcm-at-zama) [reply conftest.py:215]: - it's not new, it was already there
- and what's the problem with seeding: I want to be sure of the conditions in the `(f_i, c_j, t_k)` test, so I force the seeds here

File: deps_licenses/licenses_mac_intel_user.txt.md5 (hunk 10 above)
Comment (bcm-at-zama) [reply deps_licenses/licenses_mac_intel_user.txt.md5:1]: the deps have not changed, I was said by @fd0r there is a racing problem

File: conftest.py (hunk 5 above)
Comment (RomanBredehoft) [reply conftest.py:176]: hmm, so you're saying running `pytest ... --randomly-seed` with nothing after won't trigger this issue ? 

if so ok but yes maybe put it more like `--randomly-seed has not been set properly` then

Comment (bcm-at-zama) [reply conftest.py:176]: no it fails, but it's already caught by the system, much before, directly in randomly-seed or pytest:

```
python -m pytest -xsvv "tests/common/test_utils.py::test_compute_bits_precision[x0-8]" --randomly-seed     
/Users/benoitchevalliermames/Documents/Zama/Git/concrete-ml/.venv/lib/python3.9/site-packages/_pytest/config/__init__.py:330: PluggyTeardownRaisedWarning: A plugin raised an exception during an old-style hookwrapper teardown.
Plugin: helpconfig, Hook: pytest_cmdline_parse
UsageError: usage: __main__.py [options] [file_or_dir] [file_or_dir] [...]
__main__.py: error: argument --randomly-seed: expected one argument
For more information see https://pluggy.readthedocs.io/en/stable/api_reference.html#pluggy.PluggyTeardownRaisedWarning
  config = pluginmanager.hook.pytest_cmdline_parse(
ERROR: usage: __main__.py [options] [file_or_dir] [file_or_dir] [...]
__main__.py: error: argument --randomly-seed: expected one argument
```

File: conftest.py (hunk 6 above)
Comment (bcm-at-zama) [reply conftest.py:183]: ok

File: conftest.py (hunk 7 above)
Comment (RomanBredehoft) [reply conftest.py:187]: ah yes I see, ok looks fine actually, I was first suspicious of this `len("concrete-ml/")` but I get it now, it's like doing some kind of regex manually

File: conftest.py (hunk 2 above)
Comment (RomanBredehoft) [reply conftest.py:191]: yes yes I got that, it's just taht your above comment says it computes  " hash(f_i, c_j, t_k, randomly-seed)" but in the code I only see 3 variables

but now I guess  `str(request.node.name)` is actually the `f_i, c_j,` part. `relative_file_path` being `t_k`

File: conftest.py (hunk 7 above)
Comment (bcm-at-zama) [reply conftest.py:187]: yes, it's possible you guys have a cleaner way to do that

File: conftest.py (hunk 6 above)
Comment (bcm-at-zama) [reply conftest.py:199]: done, ok to resolve?

File: conftest.py [hunk 11]
```diff
@@ -176,37 +168,77 @@ def function_to_seed_torch(seed):
...
 
 @pytest.fixture(autouse=True)
-def autoseeding_of_everything(record_property, request):
+def autoseeding_of_everything(request):
     """Seed everything we can, for determinism."""
-    main_seed = request.config.getoption("--forcing_random_seed", default=None)
 
-    if main_seed is None:
-        main_seed = random.randint(0, 2**64 - 1)
+    # Explanations on the seeding system:
```
Comment (bcm-at-zama) [inline conftest.py:174]: new comments @RomanBredehoft

File: conftest.py (hunk 8 above)
Comment (RomanBredehoft) [reply conftest.py:214]: well, the line I comment on : `    sub_seed = int(hash_value, 16) % 2**64`

is it because the has value (in hexa) can go above 2**64 b ? I would jut add a comment around this part (and the above lines) to add a simple `# Get the subseed by getting the hash value and retrieving an int64 value`

also : 
- for seeding I thought numpy prefered 32b values (in the code we often use % 2**31 for seed iirc)
- I guess the risk of collisiions because of this %64 are minimal, right  ?

File: conftest.py (hunk 3 above)
Comment (RomanBredehoft) [reply conftest.py:222]: okok got it you were commenting the actual workflow, not what users was supposed to do my bad !

File: conftest.py (hunk 2 above)
Comment (bcm-at-zama) [reply conftest.py:207]: yes: eg, if you launch `python -m pytest -xsvv "tests/common/test_utils.py::test_compute_bits_precision[x0-8]"`, you'll see

```
Using randomly_seed=2216371280
Using derivation_string='tests/common/test_utils.py # test_compute_bits_precision[x0-8] # 2216371280'
Using sub_seed=16202174809221803539
```

so everything is here: the relative test file, the config, the function and the randomly seed.

File: conftest.py (hunk 9 above)
Comment (RomanBredehoft) [reply conftest.py:231]: okok if it's already there all is good

File: deps_licenses/licenses_mac_intel_user.txt.md5 (hunk 10 above)
Comment (RomanBredehoft) [reply deps_licenses/licenses_mac_intel_user.txt.md5:1]: aah ok my bad yes it's because mac intel licenses were not updated last time got it !

File: conftest.py (hunk 8 above)
Comment (bcm-at-zama) [reply conftest.py:214]: - the hash is a SHA256, so it's 256b. And random.seed wants a 64b seed and numpy.random.seed wants a 32 seed; so I reduce to 64b and I have `numpy.random.seed(seed % 2**32)`
- using % 2**31 reduces to 31b, there is one bit loss here
- we have collisions on 64b in O(2**32) but it's not a problem here [what's the attack] and it's not because of us [it's python and numpy which want reduced seeds]

File: conftest.py (hunk 9 above)
Comment (bcm-at-zama) [reply conftest.py:231]: yes, it just moved by few lines

File: deps_licenses/licenses_mac_intel_user.txt.md5 (hunk 10 above)
Comment (bcm-at-zama) [reply deps_licenses/licenses_mac_intel_user.txt.md5:1]: not sure, I'll let you fix that guys, it's not related to this PR and apparently you know the problem, CC @fd0r

File: conftest.py (hunk 8 above)
Comment (RomanBredehoft) [reply conftest.py:214]: yeah got it thanks !  well I would just add a little line to simply explain that, nothing fancy

File: conftest.py (hunk 5 above)
Comment (RomanBredehoft) [reply conftest.py:190]: got it, so yeah maybe just update the message and that should be enough !

File: conftest.py (hunk 6 above)
Comment (RomanBredehoft) [reply conftest.py:199]: yes thanks !

File: conftest.py (hunk 8 above)
Comment (bcm-at-zama) [reply conftest.py:216]: Adding a line.

File: conftest.py (hunk 11 above)
Comment (RomanBredehoft) [reply conftest.py:174]: yes great thanks  !

File: conftest.py (hunk 5 above)
Comment (RomanBredehoft) [reply conftest.py:190]: I was more referring to my second suggestion : `--randomly-seed has not been set properly` as clearly it's not a user problem !

File: conftest.py (hunk 2 above)
Comment (RomanBredehoft) [reply conftest.py:207]: ok great !

File: conftest.py (hunk 8 above)
Comment (RomanBredehoft) [reply conftest.py:216]: great !

File: conftest.py (hunk 5 above)
Comment (bcm-at-zama) [reply conftest.py:190]: but the user doesn't have to set --randomly-seed
to me, it's more an internal issue, like "I arrive in an impossible situation", we should have an issue like that

Comment (RomanBredehoft) [reply conftest.py:190]: yes I know, if you prefer maybe `--randomly-seed has not been properly configured internally.`

Comment (bcm-at-zama) [reply conftest.py]: if you want: done

**Review by RomanBredehoft — DISMISSED** [review]
thanks a lot for this !

**Comment by bcm-at-zama** [comment]
Fixing conformance issues.

**Review by RomanBredehoft — APPROVED** [review]

**Review by kcelia — APPROVED** [review]
Thanks

**Comment by RomanBredehoft** [comment]
are merging this @bcm-at-zama ?

**Comment by bcm-at-zama** [comment]
Yes please, I've let you manage this guys!
